File: 2019/day02-2.ts
function solution(content: string) {
  const program = content.split(",").map((item) => parseInt(item));
  for (let noun = 0; noun < 100; noun++) {
    for (let verb = 0; verb < 100; verb++) {
      const copy = program.slice();
      copy[1] = noun;
      copy[2] = verb;
      const result = execute(copy);
      if (result === 19690720) {
        return 100 * noun + verb;
      }
    }
  }
}

function execute(program: number[]) {
  let copy = program.slice();
  let halt = false;
  let current = 0;
  while (!halt) {
    const result = executeOne(current, copy);
    current = result[0];
    copy = result[1];
    halt = result[2];
  }
  return copy[0];
}

function executeOne(
  current: number,
  program: number[]
): [number, number[], boolean] {
  const opcode = program[current];
  const copy = program.slice();
  if (opcode == 1) {
    copy[copy[current + 3]] = copy[copy[current + 1]] + copy[copy[current + 2]];
    return [current + 4, copy, false];
  }
  if (opcode == 2) {
    copy[copy[current + 3]] = copy[copy[current + 1]] * copy[copy[current + 2]];
    return [current + 4, copy, false];
  }
  if (opcode == 99) {
    return [0, copy, true];
  }
  throw Error(
    `Invalid opcode ${opcode} at ${current} in ${program.slice(
      current,
      current + 4
    )}`
  );
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2019/day02_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    program = [int(x) for x in content.split(",")]

    for noun in range(100):
        for verb in range(100):
            copy = program.copy()
            copy[1] = noun
            copy[2] = verb
            if execute(copy) == 19690720:
                return 100 * noun + verb


def execute(program: list[int]):
    current = 0
    halt = False

    while not halt:
        opcode = program[current]
        if opcode not in [1, 2, 99]:
            raise ValueError(f"Invalid opcode {opcode} at {current} in {program}")

        copy = program.copy()
        halt = False
        if opcode == 1:
            program[copy[current + 3]] = (
                copy[copy[current + 1]] + copy[copy[current + 2]]
            )
        if opcode == 2:
            program[copy[current + 3]] = (
                copy[copy[current + 1]] * copy[copy[current + 2]]
            )
        if opcode == 99:
            halt = True

        current += 4

    return program[0]


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
